### ISSUE TITLE
[Tracking] Refactor functions with many parameters to use parameter objects

### ISSUE THREAD

**Issue by neil-marcellini** [Expensify/App#50368]
**Problem:** Many functions in the App repo have a very large amount of parameters, making it very hard to work with them. If you need some data that isn’t available with existing parameters you need to add a new one. The safe way to do that is to add it as an optional param at the end of the list. In order to pass it to the function from an existing call you have to fill in all params between the last one used and this new last param. It’s time consuming to fill out the defaults, and any required params require careful considering. Re-ordering parameters is risky and requires updating all existing calls. Also, we often have cases where we have a transaction object in the component, then call a function passing various fields of the transaction split out into n parameters. Wouldn’t it be so much easier to pass the original object? Sometimes we also pass the ID of the object in Onyx only to retrieve it from an Onyx connection later.

**Solution:** Gradually refactor functions to take in parameter objects, where it’s easier to re-order. Strive to minimize the number of fields. Group parameters into sensible sub-objects. Avoid duplicate prefixes such as `parameters.chat.chatReport`, use `parameters.chat.report` instead. Thrive :smile:

[Original Slack thread.](https://expensify.slack.com/archives/C01GTK53T8Q/p1727820394837439)

To do list

- [x] Refactor one function as an example. Maybe requestMoney or something that's less critical but has a similarly large number of parameters.
- [x] Update the style guide
- [x] Create refactoring issues
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021846632713060454080</li>
        <li>Upwork Job ID: 1846632713060454080</li>
        <li>Last Price Increase: 2024-10-16</li>
<li>Automatic offers: </li>
<ul>
<li>dukenv0307 | Reviewer | 104452249</li>
<li>ChavdaSachin | Contributor | 104452250</li>
<li>mkzie2 | Contributor | 104791696</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @neil-marcellini</details>

**Comment by ChavdaSachin** [comment]
@neil-marcellini we could start with IOU.getTrackExpenseInformation, IOU.shareTrackedExpense, IOU.categorizeTrackedExpense after these three IOU.trackExpense.
These functions mostly share common data and could be refactored with ease for a start.
I would like to work on this and following if you want external contributor to work here.

**Comment by neil-marcellini** [comment]
Sure @ChavdaSachin, please go ahead and start with only IOU.categorizeTrackedExpense

**Comment by ChavdaSachin** [comment]
I am OOO for 2 days on an emergency business , will start working as soon as I am back.

**Comment by neil-marcellini** [comment]
How's it coming along?

**Comment by neil-marcellini** [comment]
Hi @ChavdaSachin. I appreciate your desire to tackle it all at once, but it's a lot easier for everyone involved if we do it in pieces. It will be faster to get the PR up for review and get it approved. I rather have many small PRs than one big one. Could you please split up your existing work into several PRs?

**Comment by mkzie2** [comment]
> I rather have many small PRs than one big one

@neil-marcellini  I'd like to work on one of the PRs if we create other issues.

**Comment by ChavdaSachin** [comment]
hey @neil-marcellini I have my code prepared for `IOU.categorizeTrackedExpense` but I am unable to test it.
There seems to be a BE issue related to categorizeTrackedExpanse when `tax Rates` feature is enabled for the workspace.

<details><summary>BE response</summary>
```
{
    "jsonCode": 402,
    "message": "Invalid ReportID '0' for the report store",
    "requestID": "8decfcc139663cf4-BOM",
    "onyxData": []
}
```
</details>

<details><summary>Parameter passed</summary>
```
{
    "policyID": "95B63EE2BC4F1E51",
    "transactionID": "160923194790882510",
    "moneyRequestPreviewReportActionID": "8261683015292608374",
    "moneyRequestReportID": "2244080037308526",
    "moneyRequestCreatedReportActionID": "3617745669044213628",
    "actionableWhisperReportActionID": "1030875335587309936",
    "modifiedExpenseReportActionID": "4765574918611489569",
    "reportPreviewReportActionID": "9087019233556138435",
    "amount": 1100,
    "currency": "INR",
    "comment": "",
    "merchant": "mock",
    "category": "Equipment",
    "tag": "",
    "taxCode": "id_TAX_EXEMPT",
    "taxAmount": 0,
    "billable": false,
    "created": "2024-11-07",
    "receipt": {}
}
```
</details>

**Comment by neil-marcellini** [comment]
Ok interesting. If that error also occurs on main you can ignore it. Can you test on a workspace without tax rates? Please raise a draft PR.

**Comment by neil-marcellini** [comment]
> > I rather have many small PRs than one big one
> 
> @neil-marcellini I'd like to work on one of the PRs if we create other issues.

Ok thank @mkzie2. Please pick a function unrelated to `IOU.categorizeTrackedExpense` and go ahead and refactor it.

**Comment by ChavdaSachin** [comment]
Draft is up, please take a look...

**Comment by mkzie2** [comment]
@dukenv0307 The [PR](https://github.com/Expensify/App/pull/52231) for `requestMoney` function is ready.

**Comment by mkzie2** [comment]
@neil-marcellini Next I think we can continue refactoring the related function with `requestMoney`. I think the next we can do is `getMoneyRequestInformation` and then `buildOnyxDataForMoneyRequest`. I can continue working on this.

**Comment by mkzie2** [comment]
@neil-marcellini Any update here?

**Comment by neil-marcellini** [comment]
Sorry for the delay I was out last week. Here's an issue for the next step. Please post your proposal there, I'll review and likely assign you. [Refactor getMoneyRequestInformation to use a parameter object](https://github.com/Expensify/App/issues/53416#top)

**Comment by neil-marcellini** [comment]
@ChavdaSachin as I said before, can you please test without tax rates to work around the error if it also occurs on main? Please report that as a separate bug. Let's keep the task of fixing the backend separate from this issue and keep refactoring functions that we are not blocked on.

Please suggest a function to refactor and I'll create an issue for you. Up next @mkzie2 will be working on [Refactor buildOnyxDataForMoneyRequest to use a parameter object](https://github.com/Expensify/App/issues/53593#top). Please comment your proposal there so I can assign you.

**Comment by ChavdaSachin** [comment]
@neil-marcellini I retested it, and BE issue seems to be resolved.
[This](https://github.com/Expensify/App/pull/52221) PR will be ready for review in an hour.

And next function I would like to work on is IOU.shareTrackedExpense.

**Comment by ChavdaSachin** [comment]
https://github.com/Expensify/App/pull/52221 
PR Ready for review.
cc. @neil-marcellini @dukenv0307

**Comment by mkzie2** [comment]
@neil-marcellini `buildOnyxDataForMoneyRequest` is done. The next I think we can do with `createDistanceRequest` (22 parameters).

**Comment by neil-marcellini** [comment]
Sweet, here's the issue [Refactor createDistanceRequest to use a parameter object](https://github.com/Expensify/App/issues/53789#top).

**Comment by Shahidullah-Muffakir** [comment]
Hi @neil-marcellini,

I see others are working on migrations. Is there another function I can help refactor to use a parameter object? Let me know!

Thanks!

**Comment by ChavdaSachin** [comment]
@neil-marcellini next I could work on IOU.shareTrackedExpense function, could you creat new issue?
Thank you

**Comment by ChavdaSachin** [comment]
@neil-marcellini bump on above^

**Comment by OfstadC** [comment]
# Payment Summary

- @ChavdaSachin paid $250 via Upwork
- @dukenv0307 paid $250 via Upwork
- @mkzie2 paid $250 via Upwork

**Comment by neil-marcellini** [comment]
Re-opening because this is a tracking issue for the project. Let's handle payment going forward in the separate sub-issues.

**Comment by neil-marcellini** [comment]
Hi, I created these two new issues and tagged you guys there. Please comment so I can assign. @dukenv0307 moving forward I'm going to let C+ get automatically assigned, because these PRs tend to be easier reviews and I think it's fair to spread those among other C+ now.

[Refactor buildOptimisticTransaction to use a parameter object](https://github.com/Expensify/App/issues/54464#top)
[Refactor shareTrackedExpense to use a parameter object](https://github.com/Expensify/App/issues/54466#top)

**Comment by neil-marcellini** [comment]
I marked @CortneyOfstad as the owner and I'll DM her to get to this when she comes back to work at the end of this week, or early next week.

**Comment by OfstadC** [comment]
@neil-marcellini I think i'm a little lost here 😅 

So i've paid $250 [here](https://github.com/Expensify/App/issues/50368#issuecomment-2557299720). 

What else needs to be paid here - another $250 or $500? Just to @dukenv0307 or others as well? 

Do we have separate issues for these payments or should I create those?

**Comment by OfstadC** [comment]
Working on paying @dukenv0307 another $250. If I missed anyone please let me know!

**Comment by OfstadC** [comment]
## Further Payment Summary 
@dukenv0307 has been [paid](https://www.upwork.com/nx/wm/workroom/39421323/overview) $250 via Upwork 
_______________________
So I think that's it for payments here. If I need to action anything else, please let me know once you return from OoO @neil-marcellini 😃 

I'll leave this open since this is an ongoing issue

**Comment by neil-marcellini** [comment]
@mkzie2 hi yes I'll create an issue for trackExpense

**Comment by neil-marcellini** [comment]
@mkzie2 @abdulrahuman5196 can you please give an update on the trackExpense issue?

**Comment by neil-marcellini** [comment]
Sure!

**Comment by neil-marcellini** [comment]
There you go @mkzie2. Please post your proposal there [[$250] Refactor splitBill to use a parameter object](https://github.com/Expensify/App/issues/56773#top)

**Comment by neil-marcellini** [comment]
Hey @Krishna2323 thanks for writing that up! I will create a bunch of issues for those. I think our guideline should be that we consider refactoring to use a parameter object at 5 parameters, and require it at 10 parameters. So I'll create issues only for those with 10+ now.

I'm also going to ask for some feedback in Slack about how we want to update the style guide and work on getting that done soon.

**Comment by samranahm** [comment]
@neil-marcellini Since, we've almost completed the Refectoring of the functions with 10+ params, For reference here are some functions that contain 7 or more parameters.

1. [`getRateForDisplay`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/DistanceRequestUtils.ts#L148-L155) (7 params)
2. [`getDistanceMerchant`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/DistanceRequestUtils.ts#L220-L227) (7 params)
3. [`updateIOUOwnerAndTotal`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/IOUUtils.ts#L73-L80) (7 params)
4. [`getIcons`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L2641-L2648) (7 params)
5. [`getReportPreviewMessage`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L3895-L3902) (7 params)
6. [`buildOptimisticTaskCommentReportAction`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L4994-L5001) (7 params)
7. [`buildOptimisticIOUReport`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L5073-L5080) (7 params)
8. [`buildOptimisticExpenseReport`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L5227-L5234) (7 params)
9. [`getIOUReportActionMessage`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L5405) (7 params)
10. [`buildOptimisticTaskReport`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L6689-L6696) (7 params)
11. [`getTaskAssigneeChatOnyxData`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/ReportUtils.ts#L8010-L8017) (7 params)
12. [`getDeleteTrackExpenseInformation`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L2252-L2259) (7 params)
13. [`getSendInvoiceInformation`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L2474-L2483) (9 params)
14. [`getUpdateMoneyRequestParams`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L3332-L3340) (8 params)
15. [`updateMoneyRequestTag`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L3955-L3962) (7 params)
16. [`updateMoneyRequestCategory`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L4087-L4094) (7 params)
17. [`updateMoneyRequestDistanceRate`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L4129-L4137) (8 params)
18. [`sendInvoice`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L4716-L4725)(9 params)
19. [`getSendMoneyParams`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L7154-L7161) (7 params)
20. [`canIOUBePaid`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/IOU.ts#L8108-L8116) (8 params)
21. [`createTaskAndNavigate`](https://github.com/Expensify/App/blob/6410259e9096d4026f1c3e45b2c0c94b81223d20/src/libs/actions/Task.ts#L116-L124) (8 params)

**Comment by neil-marcellini** [comment]
Thanks, I'll create issues for those today

**Comment by neil-marcellini** [comment]
We're ticking away at various sub issues

**Comment by neil-marcellini** [comment]
Same update

**Comment by neil-marcellini** [comment]
I think this is basically all done now. Thanks to everyone who helped out!
